### PR TITLE
[extension/healthcheckv2] Add event aggregation logic

### PR DESCRIPTION
**Pull request by mwear**
**Description:** 
This PR is the second in a series to decompose https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30673 into more manageable pieces for review. 

**Aggregator**

This PR introduces an aggregator data structure and event aggregation logic for status events. The extension implements the `StatusWatcher` optional interface, which the collector will call with a `component.StatusEvent` for each change in component status. These events will be aggregated by an aggregation function, and stored in the aggregator.

The aggregator is a recursive data structure. At the top it contains the overall status of the collector. At the next level, it contains the statuses for each pipeline, and at the level below that, it contains the statuses for each component in a pipeline. Each node in the data structure is an aggregation the status events in the level below. The overall collector status is the aggregation of the pipeline statuses, and at the next level, the pipeline statuses are the aggregations of the component statuses. The data structure allows you to query the status of the collector overall, or for individual pipelines by name. There is also a pub/sub mechanism used for streaming aggregated statuses.

**Aggregation Function**

The purpose of the aggregator is to aggregate events so that the most relevant status event bubbles to the top. This allows us to get the status of the collector overall or a pipeline through a simple lookup. There is an aggregation function that determines the priority of events and how they should be aggregated. In many cases, the result will be an existing status event. In some cases a new event will be synthesized. In order to match the behavior existing healthcheck extension, lifecycle events (e.g. starting, stopping, etc) are prioritized over runtime events. Next, error statuses are prioritized with PermanentErrors as higher priority than RecoverableErrors, but this can vary based on user provided configuration. If PermanentErrors are ignored by configuration, but RecoverableErrors are included, then RecoverableErrors will take priority over PermanentErrors.

The StatusWatcher interface receives immutable events of type `component.StatusEvent`. Since we sometimes need to synthesize new events during aggregation, an `Event` interface was introduced so that the aggregator can use `component.StatusEvent` instances or instances of events synthesized by the status package.

It's worth mentioning that there is [existing status event aggregation logic](https://github.com/open-telemetry/opentelemetry-collector/blob/main/component/status.go#L101-L190) in collector core, but it did not meet the needs of this extension. It does not prioritize lifecycle events over error events, and it will always prioritize permanent errors over recoverable. By prioritizing lifecycle events over error events we can return a 503 when restarting a collector rather than a 500 when a collector in a final state, such as PermanentError. This is necessary to match the behavior of the existing extension. Since users have the option to include or ignore recoverable and permanent errors, we need the ability to prioritize them accordingly. We can discuss what the fate of the aggregation code in core should be.

**Examples**
Below are examples of overall collector and pipeline status that are based on the aggregator data structure. The rendering of the examples will come in a later PR. You can also look at the parent PR to see how all of this fits together. Note that the pipeline status example is a subtree of the overall collector status.

Overall collector status:

```json
{
    "start_time": "2024-01-18T17:27:12.570394-08:00",
    "healthy": true,
    "status": "StatusRecoverableError",
    "error": "rpc error: code = ResourceExhausted desc = resource exhausted",
    "status_time": "2024-01-18T17:27:32.572301-08:00",
    "components": {
        "extensions": {
            "healthy": true,
            "status": "StatusOK",
            "status_time": "2024-01-18T17:27:12.570428-08:00",
            "components": {
                "extension:healthcheckv2": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.570428-08:00"
                }
            }
        },
        "pipeline:metrics/grpc": {
            "healthy": true,
            "status": "StatusRecoverableError",
            "error": "rpc error: code = ResourceExhausted desc = resource exhausted",
            "status_time": "2024-01-18T17:27:32.572301-08:00",
            "components": {
                "exporter:otlp/staging": {
                    "healthy": true,
                    "status": "StatusRecoverableError",
                    "error": "rpc error: code = ResourceExhausted desc = resource exhausted",
                    "status_time": "2024-01-18T17:27:32.572301-08:00"
                },
                "processor:batch": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571132-08:00"
                },
                "receiver:otlp": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571576-08:00"
                }
            }
        },
        "pipeline:traces/http": {
            "healthy": true,
            "status": "StatusOK",
            "status_time": "2024-01-18T17:27:12.571625-08:00",
            "components": {
                "exporter:otlphttp/staging": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571615-08:00"
                },
                "processor:batch": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571621-08:00"
                },
                "receiver:otlp": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571625-08:00"
                }
            }
        }
    }
}
```
Status for pipeline `traces/http`:

```json
{
    "start_time": "2024-01-18T17:27:12.570394-08:00",
    "healthy": true,
    "status": "StatusOK",
    "status_time": "2024-01-18T17:27:12.571625-08:00",
    "components": {
        "exporter:otlphttp/staging": {
            "healthy": true,
            "status": "StatusOK",
            "status_time": "2024-01-18T17:27:12.571615-08:00"
        },
        "processor:batch": {
            "healthy": true,
            "status": "StatusOK",
            "status_time": "2024-01-18T17:27:12.571621-08:00"
        },
        "receiver:otlp": {
            "healthy": true,
            "status": "StatusOK",
            "status_time": "2024-01-18T17:27:12.571625-08:00"
        }
    }
}
```

**Link to tracking Issue:** #26661

**Testing:** Units / manual

**Documentation:** Comments, etc

### CODE REVIEW

File: extension/healthcheckv2extension/internal/status/aggregation.go
Comment (jpkrohling) [inline extension/healthcheckv2extension/internal/status/aggregation.go:52]: is this priority going to be fixed for the lifecycle of the component?

File: extension/healthcheckv2extension/internal/status/aggregation.go
Comment (jpkrohling) [inline extension/healthcheckv2extension/internal/status/aggregation.go:126]: can you add some comments/tests for this specifically? I assume this is to return the earliest error, but would be nice to see this described more properly

File: extension/healthcheckv2extension/internal/status/aggregation.go
Comment (jpkrohling) [inline extension/healthcheckv2extension/internal/status/aggregation.go:131]: This got me: matchingEvent is happening later, and if it's not an error, you are getting the latest occurrence of the event, not earliest. Why? Can you add a comment or a test?

File: extension/healthcheckv2extension/internal/status/aggregator.go
Comment (jpkrohling) [inline extension/healthcheckv2extension/internal/status/aggregator.go:110]: RLock/RUnlock? I think you are only reading values here, right? Or is the clone operation changing `a` as well?

File: extension/healthcheckv2extension/internal/status/aggregator.go
Comment (jpkrohling) [inline extension/healthcheckv2extension/internal/status/aggregator.go:87]: To confirm: this is intended to guard both `aggregateStatus` and `subscriptions`, so that no concurrent updates to subscriptions and aggregateStatus can happen, right?

File: extension/healthcheckv2extension/internal/status/package_test.go
Comment (jpkrohling) [inline extension/healthcheckv2extension/internal/status/package_test.go:1]: wasn't this auto-generated?

File: extension/healthcheckv2extension/internal/status/aggregation.go
Comment (mwear) [reply extension/healthcheckv2extension/internal/status/aggregation.go:112]: I replaced the two inline funcs you mentioned with a conditional. I can probably take this one step further and make the `aggregationFunc` a method on the `Aggregator`. Part of me prefers this logic to stand on its own, since it's somewhat complicated, but I'm open to whatever makes it the most readable / understandable / maintainable etc.

File: extension/healthcheckv2extension/internal/status/aggregation.go
Comment (mwear) [reply extension/healthcheckv2extension/internal/status/aggregation.go:61]: Yes, this is actually fixed for the aggregator as a whole (e.g. all components) for the lifetime of the collector process.

File: extension/healthcheckv2extension/internal/status/aggregator.go
Comment (mwear) [reply extension/healthcheckv2extension/internal/status/aggregator.go:87]: Correct

File: extension/healthcheckv2extension/internal/status/aggregation.go
Comment (mwear) [reply extension/healthcheckv2extension/internal/status/aggregation.go:133]: This was partially described in the comment above `newAggregationFunc`, but I expanded it, added some inline comments, and a test. 

We are interested in the earliest error event for two reasons: causality and recoverable errors.  In the event that a component error cascades to other components, the earliest is likely to be the cause. We expect recoverable errors to recover within their recovery duration and we don't want later recoverable errors to shadow earlier ones in the aggregate status. For the non-error case, we use the latest event as it represents the last time a successful status was reported. tl;dr for an error, we're interested in when it started, otherwise we want to know the latest successful status.

File: extension/healthcheckv2extension/internal/status/package_test.go
Comment (mwear) [reply extension/healthcheckv2extension/internal/status/package_test.go:1]: It wasn't. I deleteed it and reran `make generate` to double check and it only seems to generate a package_test for the top level component directory.

File: extension/healthcheckv2extension/internal/status/aggregation.go
Comment (jpkrohling) [reply extension/healthcheckv2extension/internal/status/aggregation.go:112]: a top-level static function would also work, but the current code looks good

**Review by jpkrohling — APPROVED** [review]

**Comment by Rutledge** [comment]
Hi there! We at Scorecard would love to see this merged. Know this has been WIP for a bit- is there an approximate timeline for it landing?

**Comment by jpkrohling** [comment]
@mwear, do you want someone else's review as well? I'm ready to hit the green button.

**Comment by mwear** [comment]
If one review is sufficient, I'm fine with that and will prepare the next PR. There is still time for people to make suggestions if they have them during the follow up PRs.

**Comment by jpkrohling** [comment]
@evan-bradley , @codeboten , if you are interested in reviewing this, let me know and I'll wait, otherwise I'll merge this by Friday morning UTC.

File: extension/healthcheckv2extension/internal/status/aggregator.go
Comment (evan-bradley) [inline extension/healthcheckv2extension/internal/status/aggregator.go:165]:
```suggestion
// A `Verbose` verbosity specifies that subtrees should be returned with the *AggregateStatus.
```

File: extension/healthcheckv2extension/internal/status/aggregator.go
Comment (evan-bradley) [inline extension/healthcheckv2extension/internal/status/aggregator.go:197]: Not totally sure what the efficiency of this is, but we may want to consider using a doubly-linked list if we expect a large number of subscribers.

The OpAMP extension [implements a similar notification pattern](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/opampextension/registry.go#L28) and uses the Go stdlib `container/list` package to try to make removals more efficient.

File: extension/healthcheckv2extension/internal/status/aggregator_test.go
Comment (evan-bradley) [inline extension/healthcheckv2extension/internal/status/aggregator_test.go:1]: Could you add a test or two for unsubscribing?

File: extension/healthcheckv2extension/internal/status/aggregator.go
Comment (evan-bradley) [inline extension/healthcheckv2extension/internal/status/aggregator.go:129]: Could we have a test case that checks for this? I think the current tests are only for everything and for pipelines.

File: extension/healthcheckv2extension/internal/status/aggregation.go
Comment (evan-bradley) [inline extension/healthcheckv2extension/internal/status/aggregation.go:105]: I think we're missing a test case to confirm that even when recoverable errors are prioritized, a permanent error is returned if a recoverable error is not present.

File: extension/healthcheckv2extension/internal/status/aggregation.go
Comment (evan-bradley) [reply extension/healthcheckv2extension/internal/status/aggregation.go:126]: I marked this unresolved because I'm not clear on this either, and I don't believe we have tests for a case where the top-level status is an error and we hit this case. I understand why we do the ordering if we hit this, but it would be nice to see more details about why this case is separate from the one below where we return `matchingEvent` if `isError` is set (on line 139 as of this comment).

**Review by evan-bradley — COMMENTED** [review]
Overall looks good to me. None of my comments are blocking.

**Comment by mwear** [comment]
Thanks for the thorough review @evan-bradley. @jpkrohling, hold off on merging until I address the feedback.

File: extension/healthcheckv2extension/internal/status/aggregation.go
Comment (mwear) [reply extension/healthcheckv2extension/internal/status/aggregation.go:79]: This was my initial instinct as well. If you look closely at the cases, they do not all conform to:

```
if _, ok := seen[s]; ok {
	return s
}
```

Specifically, there is an edge case where a component can be stopped, and at least one other component is not stopped or stopping, but the aggregate status is stopping. The placement of this clause is important also. In code it's:

```
if _, isStopped := seen[component.StatusStopped]; isStopped {
	return component.StatusStopping
}
```

I don't love the verbosity of the code and for a moment looked into introducing a `rule` type that could be put into a slice and iterated over. It seemed overkill, and probably still is. We can consider going that route if we need to support more complexity in the future. For now I am ok with this ugly, but unsurprising implementation.

File: extension/healthcheckv2extension/internal/status/aggregator.go
Comment (mwear) [reply extension/healthcheckv2extension/internal/status/aggregator.go:198]: I started down the road of using a list and it became uglier than I had hoped. When subscribing the caller gets a `chan <- *AggregateStatus`, and this is what is used to unsubscribe. Internally, the subscription is a struct with a channel and verbosity. I'd prefer not to return that to the caller, because only the channel is of any use. The net of this, is that you need to iterate through all of the subscriptions to find one with the matching channel. In order to make this work, I'd have to introduce two lookup maps. One that maps the channel to its subscription, and one that maps the subscription to the list containing it. It can be done, but this might be a premature optimization. We can certainly do this if it's needed in the future. Is that reasonable or am I missing something that makes this easier?

Comment (mwear) [reply extension/healthcheckv2extension/internal/status/aggregator.go:198]: For additional context, this functionality is used for gRPC streaming healthchecks only. I do not expect an overly large number of clients.

Comment (mwear) [reply extension/healthcheckv2extension/internal/status/aggregator.go]: I took a second look at this and updated subscribe to return a channel and a function to unsubscribe. The UnsubscribeFunc cleared up the issues I was running into and this now uses container/list for subscriptions.

File: extension/healthcheckv2extension/internal/status/aggregation.go
Comment (evan-bradley) [reply extension/healthcheckv2extension/internal/status/aggregation.go:79]: That's fair, I don't think we would gain much right now with anything like a `rule` type. Thanks for the details, let's stick with the current implementation.

**Review by evan-bradley — APPROVED** [review]

**Comment by mwear** [comment]
Is this ready to merge? If not, let me know what needs to be addressed. If it is ready, I can start prepping the next PR.

**Comment by evan-bradley** [comment]
I was waiting for any last comments to trickle in, but at this point I think we're okay. We can address anything else in follow-ups.

**Comment by evan-bradley** [comment]
Thanks again for spearheading this effort @mwear.

**Pull request merged**
